src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,504 @@
+            //      for i = 0 to ...
+            //
+            // We can't convert this as it would change important semantics.
+            // NOTE: we could potentially update this if we saw that the variable was not used
+            // after the for-loop.  But, for now, we'll just be conservative and assume this means
+            // the user wanted the 'i' for some other purpose and we should keep things as is.
+            var iterationSymbol = semanticModel.GetSymbolInfo(iterationVariable.Parent, cancellationToken).GetAnySymbol();
+            if (iterationSymbol != null)
+            {
+                if (iterationSymbol.Locations.Length != 1 ||
+                    !iterationSymbol.Locations[0].IsInSource ||
+                    iterationVariable != iterationSymbol.Locations[0].FindToken(cancellationToken))
+                {
+                    // was a reference to some other variable.
+                    return;
+                }
+            }
+
+            // Make sure we're starting at 0.
+            var initializerValue = semanticModel.GetConstantValue(initializer, cancellationToken);
+            if (!(initializerValue.HasValue && initializerValue.Value is 0))
+            {
+                return;
+            }
+
+            // Make sure we're incrementing by 1.
+            if (stepValueExpressionOpt != null)
+            {
+                var stepValue = semanticModel.GetConstantValue(stepValueExpressionOpt);
+                if (!(stepValue.HasValue && stepValue.Value is 1))
+                {
+                    return;
+                }
+            }
+
+            var collectionType = semanticModel.GetTypeInfo(collectionExpression, cancellationToken);
+            if (collectionType.Type == null && collectionType.Type.TypeKind == TypeKind.Error)
+            {
+                return;
+            }
+
+            var containingType = semanticModel.GetEnclosingNamedType(context.Span.Start, cancellationToken);
+            if (containingType == null)
+            {
+                return;
+            }
+
+            var ienumerableType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var ienumeratorType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T);
+
+            // make sure the collection can be iterated.
+            if (!TryGetIterationElementType(
+                    containingType, collectionType.Type,
+                    ienumerableType, ienumeratorType,
+                    out var iterationType))
+            {
+                return;
+            }
+
+            // If the user uses the iteration variable for any other reason, we can't convert this.
+            var bodyStatements = GetBodyStatements(forStatement);
+            foreach (var statement in bodyStatements)
+            {
+                if (iterationVariableIsUsedForMoreThanCollectionIndex(statement))
```

Comment:
Do we handle the case where the collection expression is re-assigned in the loop body and bail out?

Reply:
Also how about the case where we invoke methods on the collection such as Add, Remove, etc. within the loop? We should detect and bail out otherwise the foreach enumeration will throw exception for underlying collection modified during loop iteration.

Reply:
> Also how about the case where we invoke methods on the collection such as Add, Remove, etc. within the loop? We should detect and bail out otherwise the foreach enumeration will throw exception for underlying collection modified during loop iteration.

No.  We can choose to detect this.  But we should not bail out.  We shoudl give a warning that semantics may change due to the collectoin changing.  Again, as above, this is not an analyzer/suggestion. We are not suggesting to the user "you should change this to iteration".  Instead, the user is coming and saying: i wrote this with indexing, now i'd like it to be enumeration, can you speed that up for me?"  We'd like to say 'yes' as much as possible, albeit letting them know when there may be unforseen consequences.  I'll add checks for invocations and reassignment and add warnings.